### PR TITLE
simple command costmap api - first few functions

### PR DESCRIPTION
**Pull request by stevedanomodolor**
Signed-off-by: Stevedan Ogochukwu Omodolor <stevedan.o.omodolor@gmail.com>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (add tickets here #1) |
| Primary OS tested on | (Ubuntu) |
| Robotic platform tested on | (gazebo ) |

---

## Description of contribution in a few bullet points
* Add a costmap API for OccupancyGrid messages coming into the Simple Commander API
* Constructor for OccupancyGrid which stores the relevant information
* Implementation of the  following first few functions
  * getSizeInCellsX / getSizeInCellsY
  * getSizeInMetersX / getSizeInMetersY
  *  getOriginX / getOriginY
  * getResolution
  * getGlobalFrameID
  * getCostmapTimestamp

## Description of documentation updates required from your changes
* Added new api, documentation on this api might be required in the simple commander page documentation. 

---

## Future work that may be required in bullet points
* Implementation of the following functions 
  *  getCost at map pixels and index
  * setCost at map pixels
  * mapToWorld
  * worldToMap
  * getIndex

## Related issue 
* https://github.com/ros-planning/navigation2/issues/3151
#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    def getSizeInMetersX(self):
+        ...
+    def getOriginX(self):
+        ...
+    def getOriginY(self):
+        ...
+    def getResolution(self):
+        ...
+    def getGlobalFrameID(self):
+        ...
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:26]: The header stores the frame_id and the timestamp

**Review by SteveMacenski — COMMENTED** [review]
LGTM, just needs the functions implemented!

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py (hunk 1 above)
Comment (stevedanomodolor) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:26]: Thank you @SteveMacenski . I will include that.

**Comment by stevedanomodolor** [comment]
@SteveMacenski Task completed

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#! /usr/bin/env python3
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:1]: Please lint with `ament_flake8`

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# limitations under the License.
+
+import numpy as np
+
+
+class PyCostmap2D:
+    """
+    PyCostmap2D.
+
+    Costmap API for OccupancyGrid messages coming into the Simple Commander API
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:24]:
```suggestion
    Costmap Python3 API for OccupancyGrids to populate from published messages
```

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    """
+
+    def __init__(self, occupancy_map):
+        self.size_x_ = occupancy_map.info.width
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:28]: I don't think in python we typically do trailing underscores - isn't `self.` already an indicator of a member variable?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
LGTM.

It would be good to add a new section to https://navigation.ros.org/commander_api/index.html for "Costmap API" with the same kind of table from the "Commander API" section with a short paragraph at the top explaining:
- This is a Python3 API for costmap 2d messages from the stack
- Provides the basic conversion, get/set, and handling semantics found in the costmap 2d C++ API

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py (hunk 4 above)
Comment (stevedanomodolor) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:28]: Its an old habit I picked up from cpp😅, I can change it to the other convention if needed. In theory as you said, self do indeed indicate that its part of the class.

**Comment by stevedanomodolor** [comment]
> LGTM.
> 
> It would be good to add a new section to https://navigation.ros.org/commander_api/index.html for "Costmap API" with the same kind of table from the "Commander API" section with a short paragraph at the top explaining:
> 
> * This is a Python3 API for costmap 2d messages from the stack
> * Provides the basic conversion, get/set, and handling semantics found in the costmap 2d C++ API

That sound good. I will make the changes as soon as possible

**Comment by stevedanomodolor** [comment]
> * This is a Python3 API for costmap 2d messages from the stack
> * Provides the basic conversion, get/set, and handling semantics found in the costmap 2d C++ API

Related pull request for documentation. 
* https://github.com/ros-planning/navigation.ros.org/pull/342

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@stevedanomodolor would you have interest in working on a simplified `rclpy` API with me? I'm working on a currently private repository to create a simplified ROS 1-like API for both python and C++. I've gotten started with python first since I think its easier to quickly jot something up for it and then translate to C++ once finished. 

I'm just getting started with it today. I could use some help completing the API and writing tests to make sure everything works as expected.  Figure I have no particular need to do everything alone if there's interest!

**Comment by stevedanomodolor** [comment]
@SteveMacenski Yes, I would love to

**Comment by SteveMacenski** [comment]
Sweet, I added you to the repo but you should join the nav2 slack (invite link in readme https://github.com/ros-planning/navigation2) so we can chat about what needs to be done / do. But I think the bullet on creating the timer/time/rate/duration simplified wrapper classes would be a good starting point. The tl;dr there is (1) remove callback groups / executor / QoS settings from the constructor and use (2) use double seconds for all APIs involving time (3) try to mimick roughly speaking the ROS 1 API where reasonable.

You can see some examples in the rclpy simple repo already, but you don't need to reimplement these things from scratch, just wrap about the rclpy content with simplified APIs for external users that sets default QoS / etc so the user doesn't have to think about it or know anything about it. If you look at the readme, I give a decent overview of the intention

**Pull request merged**